### PR TITLE
Fix NPC actiosn again

### PR DESCRIPTION
**Pull request by warp-core**
**Bugfix:** Alternative to #9265.

## Fix Details
Maps NPC action triggers to event masks. In order to trigger the corresponding action, the ship events that have occured to every ship in that NPC must return true when performing a bitwise AND with that mask.
In the case of the KILL trigger, this requires that every ship in that NPC has had the capture or destroy event occur to it.
This fixes an issue where the KILL trigger was not working correctly if the last ship in the NPC to be killed was captured but not all the others were. Since the capture event was sent, all other ships were required to have had the capture event in order to trigger the kill action, resulting in it not triggering if some were outright destroyed. This is not desired.

## Testing Done
Use the following mission and perform different combinations of destroying and capturing the two derelict Sparrows.

mission "test"
	job
	repeat
	npc kill
		personality derelict
		government Pirate
		ship "Sparrow" "Capture"
		ship "Sparrow" "Kill"
		on kill
			dialog `Killed`
		on destroy
			dialog `Destroyed`
		on capture
			dialog `Captured`

Get the following results. Old being current master, new being this PR:

Events | old | new | desired
-- | -- | -- | --
Both destroyed | "Killed" "Destroyed" | "Killed" "Destroyed" | "Killed" "Destroyed"
Both captured | "Killed" "Captured" | "Killed" "Captured" | "Killed" "Captured"
First destroyed, second captured | Nothing | "Killed" | "Killed"
First captured, second destroyed | "Killed" "Destroyed" | "Killed" | "Killed"

### CODE REVIEW

File: source/NPC.cpp [hunk 1]
```diff
@@ -770,14 +770,36 @@ void NPC::DoActions(const ShipEvent &event, bool newEvent, PlayerInfo &player, U
...
+		// Some triggers cannot be met if any of the ships in this NPC have certain events.
+		// If any of the ships were captured, the DESTROY trigger will not run.
+		static const map<Trigger, int> triggerExclusions = {
+			{Trigger::DESTROY, ShipEvent::CAPTURE}
+		};
+
+		const auto requiredIt = triggerRequirements.find(trigger);
+		const int requiredEvents = requiredIt == triggerRequirements.end() ? 0 : requiredIt->second;
+		const auto excludedIt = triggerExclusions.find(trigger);
+		const int excludedEvents = excludedIt == triggerExclusions.end() ? 0 : excludedIt->second;
```
Comment (RisingLeaf) [inline source/NPC.cpp:794]:
```suggestion
		const int excludedEvents = excludedIt == triggerExclusions.end() ? 0 : excludedIt->second;

```
I think it makes the code better to read with an empty line before comments

File: source/NPC.cpp [hunk 2]
```diff
@@ -770,14 +770,36 @@ void NPC::DoActions(const ShipEvent &event, bool newEvent, PlayerInfo &player, U
...
+			{Trigger::SCAN_OUTFITS, ShipEvent::SCAN_OUTFITS},
+			{Trigger::PROVOKE, ShipEvent::PROVOKE},
+			{Trigger::DISABLE, ShipEvent::DISABLE},
+			{Trigger::BOARD, ShipEvent::BOARD},
+			{Trigger::CAPTURE, ShipEvent::CAPTURE},
+			{Trigger::DESTROY, ShipEvent::DESTROY},
+			{Trigger::KILL, ShipEvent::CAPTURE | ShipEvent::DESTROY}
+		};
+
+		// Some triggers cannot be met if any of the ships in this NPC have certain events.
```
Comment (RisingLeaf) [inline source/NPC.cpp:785]:
```suggestion
		// Some Triggers cannot be met if any of the ships in this NPC have certain events.
```
see comments below, Trigger gets capitalized (this is very important 😆 )

**Review by RisingLeaf — APPROVED** [review]
looks good

File: source/NPC.cpp [hunk 3]
```diff
@@ -770,14 +770,36 @@ void NPC::DoActions(const ShipEvent &event, bool newEvent, PlayerInfo &player, U
...
+		const int excludedEvents = excludedIt == triggerExclusions.end() ? 0 : excludedIt->second;
 		// The PROVOKE Trigger only requires a single ship to receive the
 		// event in order to run. All other Triggers require that all ships
 		// be affected.
 		if(trigger == Trigger::PROVOKE || all_of(ships.begin(), ships.end(),
 				[&](const shared_ptr<Ship> &ship) -> bool
 				{
 					auto it = shipEvents.find(ship.get());
-					return it != shipEvents.end() && it->second & type;
+					return it != shipEvents.end() && it->second & requiredEvents && !(it->second & excludedEvents);
```
Comment (Amazinite) [inline source/NPC.cpp:802]:
```suggestion
					return it != shipEvents.end() && (it->second & requiredEvents) && !(it->second & excludedEvents);
```

**Pull request merged**